restorer.go

```diff
@@ -31,8 +31,10 @@ func (r *Restorer) Restore(cache Cache) error {
 		return err
 	}
 
+	var buildpackLayersToSha BuildpackLayersToSha
 	if r.restoresLayerMetadata() {
-		if err := r.LayerMetadataRestorer.Restore(r.Buildpacks, r.LayersMetadata, cacheMeta); err != nil {
+		var err error
+		if buildpackLayersToSha, err = r.LayerMetadataRestorer.Restore(r.Buildpacks, r.LayersMetadata, cacheMeta, false); err != nil {
```

Comment:
I decided to add the BuildpackLayersToSha return value to the `LayerMetadataRestorer.Restore` function, although it's possible to solve the issue without it.
If we would want to not add this return value, we would need to add a few `if-else` conditions to this function as appear [here](https://github.com/buildpacks/lifecycle/blob/main/layer_metadata_restorer.go#L72-L116) in order to decide whether to read the sha from the app layer, the cache layer, or to not read it from anywhere.
Since we call the `restoreLayerMetadata` function anyway in this flow, I decided to just use what we calculate there.

If anyone has different opinion or a better solution for it, I would love to discuss it.

Reply:
I wonder if something like this would make it so we could not return `buildpackLayersToSha` to the caller but the `LayerMetadataRestorer` could keep the data around after `Restore` and use it during `validateContents`

```
if !r.LayerMetadataRestorer.validateContents(buildpack.ID, layerName, cachedLayer.SHA) {
   r.Logger.Infof("Removing %q, wrong sha", bpLayer.Identifier())
}
```

Reply:
Interesting. But what is the advantage of doing this? Don't you think that returning a value from `Restore` is clearer?

Reply:
I see that as an implementation detail to the `LayerMetadataRestorer`. If we give the `LayerMetadataRestorer` enough information up front, I would like to think the consumers of that interface shouldn't have to see any internal data structures. I'm trying to push more data _into_ the `LayerMetadataRestorer` as opposed to have more in the `restorer` itself.

Reply:
I like the idea of not exposing internal data structures, but for code readability and keeping one flow for old and new API versions, I would replace [this line](https://github.com/buildpacks/lifecycle/pull/624/files#diff-98f352cc6ef347aa505fe6d5f3e01c3a47ee0b5d1a91d1e918343284c893b51eR88) with `r.LayerMetadataRestorer.getShaByBuildpackLayers(buildpack.ID, layerName)`.
But then I'm wondering again if it would be clearer to the reader.
Maybe even though it will be a little bit less clearer than the current code (from my point of view), it's better to do that in order to prevent exposing the `BuildpackLayersToSha` data structure.

What do you think?

Reply:
Maybe? It's hard to say without trying it out..I really would like to see this type of branching go away from `restorer.go` entirely. This is essentially platform API version branching.

```
restoreSha := !r.restoresLayerMetadata()
var layerSha string
if restoreSha {
  layerSha = data.SHA
} else {
  layerSha = buildpackLayersToSha.getShaByBuildpackLayers(buildpack.ID, layerName)
}
if layerSha != cachedLayer.SHA {
  r.Logger.Infof("Removing %q, wrong sha", bpLayer.Identifier())
  ...
}
```

In favor of something like this. 
```
if !someSmallerUnitOfWork.CacheIsValid(bpLayer, cacheLayer) {
  r.Logger.Infof("Removing %q, wrong sha", bpLayer.Identifier())
  ...
}
```

The `someSmallerUnitOfWork` here may not be `LayerMetadataRestorer`, but ideally the platform version swapping behavior can be hoisted out to where we build `restorer` and `someSmallerUnitOfWork` in the first place. But this is not a blocker - just thinking out loud really.

Reply:
I can try to do something. Let me work on it and I'll make this change in a separate PR so we can revert it if we don't like it.

Reply:
What do you think about https://github.com/buildpacks/lifecycle/pull/624/commits/a26294188114c6d1fc9f6001d4427fc0c2df2e27?
I put the new function in `LayerMetadataRestorer` in the meantime although it's not the best place for it.

Reply:
Please correct me if I'm wrong but I think that the main problem of putting the new function in `LayerMetadataRestorer` is that the `analyzer` doesn't use it.
So maybe we can create another interface for the `restorer` that will use the `LayerMetadataRestorer` interface and expand it with the new function.
Similar to `ReadWriteCloser` in the [io package](https://golang.org/pkg/io/#ReadWriteCloser).

Reply:
I like the changes! The tests are easier to read and reason about IMO. `analyzer` does have use `LayerMetadataRestorer` - did you mean something else?

Reply:
I meant that the `analyzer` doesn't use the `CacheIsValid` function.

Reply:
@yaelharel and I had a discussion offline and we agree that it's weird that `CacheIsValid` is part of `LayerMetadataRestorer`.

I think this logic:
```
	var layerSha string
	if la.useShaFile {
		data, err := layer.read(la.useShaFile)
		if err != nil {
			return false, errors.Wrapf(err, "reading layer")
		}
		layerSha = data.SHA
	} else {
		layerSha = la.buildpacksToLayersSha.getShaByBuildpackLayers(buildpackID, layer.name())
	}
```
could go into a separate service, call it `LayerSHAStore`. It could have `Add` and `Get` methods. You could optionally pass this into `LayerMetadataRestorer.Restore`. The analyzer on older platform apis could ignore it (since it doesn't need it) and the restorer on newer platform apis could use it.

So to summarize you'd have `LayerMetadataRestorer.Restore()` internally calling `LayerSHAStore.Add()` if provided by the caller and restorer calling `LayerSHAStore.Get()`. 

WDYT?

Reply:
That makes sense to me.

Reply:
Great! I think we're all on the same page.
I'll work on making these changes.

Reply:
Copying my edit for visibility:

> Edit: I just realized that restorer on older platform apis will not be able to use `LayerSHAStore.Get()` since it's not calling `LayerMetadataRestorer.Restore()`. It will have to read the layer.sha directly. I'm not sure if that makes things better or worse.

So you'd have something like 
```
restoreSha := !r.restoresLayerMetadata()
var layerSha string
if restoreSha {
  layerSha = data.SHA
} else {
  layerSha = LayerSHAStore.Get(buildpack.ID, layerName)
}
if layerSha != cachedLayer.SHA {
  r.Logger.Infof("Removing %q, wrong sha", bpLayer.Identifier())
  ...
}
```

which from https://github.com/buildpacks/lifecycle/pull/624#discussion_r643257297 we were trying to avoid.

Reply:
I just want to add that despite retaining the branching I like this option "best" for now. We could potentially explore creating a new service, say `CacheValidator`, to wrap the if/else but I think having the `LayerSHAStore` abstraction would be a convenient way to pass the data from `LayerMetadataRestorer` to `CacheValidator` anyway so we might as well go down this route and potentially improve upon. Just my two cents!

Reply:
Please see https://github.com/buildpacks/lifecycle/pull/624/commits/f584e397d9e441ee1860957da6c18e023d65a89d.